src/algebra/universal_enveloping_algebra.lean

```diff
@@ -0,0 +1,173 @@
+  map_add'  := λ a b, rfl,
+  commutes' := λ r, show _ = quot.mk _ _, by rw [algebra.smul_def'', mul_one], }
+
+/-- The natural Lie algebra morphism from a Lie algebra to its universal enveloping algebra.
+
+Note that the Lie algebra structure on the universal enveloping algebra arises from its associative
+algebra structure via the ring commutator. -/
+def ι : L →ₗ⁅R⁆ universal_enveloping_algebra R L :=
+{ to_fun    := quot.mk (rel R L) ∘ (tensor_algebra.ι R L),
+  map_add'  := λ x y, by simpa only [function.comp_app, linear_map.map_add],
+  map_smul' := λ c x, show _ = quot.mk _ _, by simp,
+  map_lie   := λ x y, quot.sound (rel.lie_compat x y), }
+
+variables (A : Type u₃) [ring A] [algebra R A] (f : L →ₗ⁅R⁆ A)
+
+/-- The universal property of the universal enveloping algebra: Lie algebra morphisms into
+associative algebras lift to associative algebra morphisms from the universal enveloping algebra. -/
+def lift : universal_enveloping_algebra R L →ₐ[R] A :=
+{ to_fun    := λ a, quot.lift_on a (tensor_algebra.lift R L (f : L →ₗ[R] A))
+    (λ a b h,
+    begin
+      induction h,
+      { -- case universal_enveloping.rel.lie_compat
+        rw [← alg_hom.to_linear_map_apply, ← linear_map.comp_apply, tensor_algebra.ι_comp_lift],
+        change f _ = _, simpa [lie_ring.of_associative_ring_bracket], },
+      { -- case universal_enveloping.rel.add_compat
+        simp only [alg_hom.map_add, add_left_inj, h_ih], },
+      { -- case universal_enveloping.rel.mul_compat_left
+        simp only [alg_hom.map_mul, h_ih], },
+      { -- case universal_enveloping.rel.mul_compat_right
+        simp only [alg_hom.map_mul, h_ih], },
+    end),
+  map_one'  := show algebra_map _ _ _ = _, by rw ring_hom.map_one,
+  map_mul'  := by { rintros ⟨⟩ ⟨⟩, change tensor_algebra.lift R L ↑f _ = _, rw alg_hom.map_mul, },
+  map_zero' := show tensor_algebra.lift R L ↑f 0 = (0 : A), by rw alg_hom.map_zero,
+  map_add'  := by { rintros ⟨⟩ ⟨⟩, change tensor_algebra.lift R L ↑f _ = _, rw alg_hom.map_add, },
+  commutes' := λ r, show tensor_algebra.lift R L ↑f _ = _, by simp [algebra.smul_def''], }
+
+@[simp] lemma ι_comp_lift' (x : L) : lift R L A f (ι R L x) = f x :=
+rfl
+
+lemma ι_comp_lift : (lift R L A f) ∘ (ι R L) = f :=
+by { ext, simp, }
+
+lemma lift_unique (g : universal_enveloping_algebra R L →ₐ[R] A) (h : g ∘ (ι R L) = f) :
+  g = lift R L A f :=
+begin
+  let φ₁ := g.comp (mk_alg_hom R L),
+  let φ₂ := (lift R L A f).comp (mk_alg_hom R L),
+  suffices : φ₁ = φ₂,
+    { ext u, rcases quot.exists_rep u with ⟨a, ha⟩, rw ← ha, change φ₁ a = φ₂ a, rw this, },
+  have h₁ : φ₁ = tensor_algebra.lift R L f,
+    { rw ← tensor_algebra.lift_unique, ext, change _ = f _, rw ← h, refl, },
+  have h₂ : φ₂ = tensor_algebra.lift R L f,
+    { rw ← tensor_algebra.lift_unique, ext, change _ = f _, rw ← ι_comp_lift R L A f, refl, },
+  rw [h₁, h₂],
+end
+
+@[ext] lemma hom_ext {g₁ g₂ : universal_enveloping_algebra R L →ₐ[R] A}
+  (h : g₁ ∘ (ι R L) = g₂ ∘ (ι R L)) : g₁ = g₂ :=
```

Comment:
I guess we could argue that we want the hypothesis to have type:
`h : (lie_algebra.of_associative_algebra_hom g₁).comp (ι R L) = (lie_algebra.of_associative_algebra_hom g₂).comp (ι R L)` but I'm dubious this is more helpful.